### DISCUSSION TITLE
how does nginx process manage go plugin server process?

### DISCUSSION THREAD

**Discussion by zffocussss** [Kong/kong#9567] [Help]
I am using go plugin server within kong.Recently the go plugin is always showing "closed".how can I debug it?
my plugin name is elastic-apm which will connect with elastic APM server and sending tracing data to it.
kong log:
*4117633 [kong] pb_rpc.lua:397 closed, context: ngx.timer,
*4117720 [kong] pb_rpc.lua:397 [elastic-apm] closed

plugin source code:
https://github.com/zffocussss/kong-elastic-apm/blob/main/main.go
kong version:
2.8.1 + pgsql db mode

**Comment by StarlightIbuki** [comment]
To answer the question in the title, one of the workers starts plugin servers in the `init_worker` phase and shut it down(in 3.0) in the `exit_worker` phase.

I need to take a look at the code snippet you provided to understand the problem. But it looks like related to a known issue that we do not wait until the plugin server is actually working to proceed to accept requests(and if the plugin server crashes it doesn't know).
